### PR TITLE
[ADR] Mother Objects

### PR DESCRIPTION
**Pull request by carlosvillu**
## Description
Create the necessary classes in @s-ui/test to start using Mother Objects in our tests.

The new classes that are exposed are:

- RequestMotherObject
- ModelMotherObject

And add the autoload feature to @s-ui/mock

### CODE REVIEW

File: packages/sui-test/src/MotherObject/FakeGenerator.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
+import {faker} from '@faker-js/faker'
+
+const IDENTITY_FN = i => i
+
+let instacenRndID
```
Comment (jordevo) [inline packages/sui-test/src/MotherObject/FakeGenerator.js:5]:
```suggestion
let instanceRndID
```

typo I guess...?

**Review by oriolpuig — COMMENTED** [review]
❤️

**Review by emiliovillu — APPROVED** [review]

File: packages/sui-mock/README.md [hunk 2]
```diff
@@ -51,6 +51,49 @@ import {getUserHandler} from './mocks/exampleGateway/user/handlers.js'
...
+  return [200, {name: 'nombre'}]
+}
+export async function post({headers, body, params, query, cookies}) {
+  return [200, {params}]
+}
+export async function put() {}
+export async function del() {}
+```
+
+Every function have to return the status code of the request and the body.
```
Comment (jordevo) [inline packages/sui-mock/README.md:95]:
```suggestion
Every function has to return the status code of the request and the body.
```

🤓 sorry!

File: packages/sui-test/src/MotherObject/Fetcher.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  static create(FetchFetcher, {InlineError}) {
+    if (_instance) return _instance
+    _instance = new MockFetcherManager(FetchFetcher, {InlineError})
+    return _instance
+  }
+
+  static restore() {
+    if (!_instance)
+      return console.warn('Unabled restore a not init MockFetcherManager')
```
Comment (jordevo) [inline packages/sui-test/src/MotherObject/Fetcher.js:14]:
```suggestion
      return console.warn('Unable to restore a non-initialized MockFetcherManager')
```

perhaps clearer?

**Review by jordevo — APPROVED** [review]

File: packages/sui-test/src/MotherObject/FakeGenerator.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+  province() {
+    return this._faker.location.county()
+  }
+
+  phone() {
+    return this._faker.phone.number()
+  }
+
+  URL() {
```
Comment (kikoruiz) [inline packages/sui-test/src/MotherObject/FakeGenerator.js:82]: For me it's strange to have here a method starting with uppercase, I'd rather use `url()`...

File: packages/sui-test/src/MotherObject/FakeGenerator.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    return this._faker.internet.url()
+  }
+
+  imgURL({removeHost} = {removeHost: true}) {
```
Comment (kikoruiz) [inline packages/sui-test/src/MotherObject/FakeGenerator.js:86]: ...and here `imgUrl()` for consistency

File: packages/sui-test/src/MotherObject/Fetcher.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        if (!self._hasInlineErrors)
+          return error
+            ? Promise.reject(error)
+            : Promise.resolve({data: response})
+
+        return Promise.resolve([error, response])
+      } else {
+        console.warn(`[MockFetcherManager#_request]
+        - Request make ${requestKey}
+        - Dont match any mock:
```
Comment (kikoruiz) [inline packages/sui-test/src/MotherObject/Fetcher.js:82]:
```suggestion
        - Doesn't match any mock:
```

File: packages/sui-test/src/MotherObject/Fetcher.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        delete self._fakeRequests[requestKey]
+        if (!self._hasInlineErrors)
+          return error
+            ? Promise.reject(error)
+            : Promise.resolve({data: response})
+
+        return Promise.resolve([error, response])
+      } else {
+        console.warn(`[MockFetcherManager#_request]
+        - Request make ${requestKey}
```
Comment (kikoruiz) [inline packages/sui-test/src/MotherObject/Fetcher.js:81]:
```suggestion
        - Request made: ${requestKey}
```

**Review by kikoruiz — APPROVED** [review]

**Comment by ferransimon** [comment]
Why is this pull request mixing MSW and MotherObjects code? Couldn't these be in two separate branches since they originate from two different ADRs?

**Pull request merged**